### PR TITLE
testsuite: refactor mtest to share code between c and cxx tests

### PR DESCRIPTION
**Pull request by hzhou**
Instead of duplicate code between `mtest.c` and `mtest.cxx`, use `mtest.inl` include file to share code.

`.inl` extension has been used in many other projects, referring to `inline`, but has become a convention for source/object level include files.

### CODE REVIEW

**Comment by hzhou** [comment]
test:jenkins/ch3/tcp

**Comment by hzhou** [comment]
test:jenkins/ch3/tcp

**Comment by hzhou** [comment]
@pavanbalaji This PR is ready for your review.

**Comment by pavanbalaji** [comment]
What's the reason for the `.inl` file?  Can't we keep that as a `.c` file and link it to both the C and C++ tests?

**Comment by hzhou** [comment]
> What's the reason for the `.inl` file? Can't we keep that as a `.c` file and link it to both the C and C++ tests?

It is a simpler. It almost works the same as before just without duplicating the code. What is the reason to make it into separate compile unit?

**Comment by pavanbalaji** [comment]
> > What's the reason for the `.inl` file? Can't we keep that as a `.c` file and link it to both the C and C++ tests?
> 
> It is a simpler. It almost works the same as before just without duplicating the code. What is the reason to make it into separate compile unit?

My reasoning was to avoid including C code, which I feel is more of a role for `.h` files.

**Comment by hzhou** [comment]
It is C++ code as far as compiler concerns :)

**Comment by pavanbalaji** [comment]
> It is C++ code as far as compiler concerns :)

I understand how the compiler sees it.  But that was not what I was saying.

**Comment by pavanbalaji** [comment]
If the concern is only about writing the build system, I can help with it.  It's not as difficult as you might think.  In fact, some would even consider that simpler, instead of the `.inl` includes.  :-)

**Comment by hzhou** [comment]
We want the code be treated by compiler in its proper C++ way, right? Otherwise, we have to worry about the difference between C and C++ compilers, even though that is being kept minimal.

Also by commenting on which is a simpler solution does not mean the same as difficulty.

**Comment by hzhou** [comment]
test:jenkins/ch3/tcp

File: test/mpi/include/mpitestcxx.h [hunk 1]
```diff
@@ -8,28 +8,27 @@
...
+#endif
+#else
+#include <iostream.h>
 #endif
-
-#include <string.h>
-/*
- * Init and finalize test
- */
 void MTest_Init(void);
```
Comment (hzhou) [inline test/mpi/include/mpitestcxx.h]: `gnuindent` is having trouble dealing here, keep concatenating `void MTest_Init(void);` to the previous comment line.

**Comment by hzhou** [comment]
test:mpich/quick

**Comment by hzhou** [comment]
test:jenkins/ch3/tcp

**Comment by hzhou** [comment]
test:jenkins/ch3/tcp

**Comment by hzhou** [comment]
@pavanbalaji The tests seems to be good. It is ready for your review.

File: test/mpi/Makefile.mtest [hunk 2]
```diff
@@ -16,20 +16,18 @@
 
 # AM_CPPFLAGS are used for C++ code as well
 AM_CPPFLAGS = -I$(top_builddir)/include -I$(top_srcdir)/include
-LDADD = $(top_builddir)/util/mtest.o
+LDADD = $(top_builddir)/util/mtest.$(OBJEXT) $(top_builddir)/util/mtest_common.$(OBJEXT)
```
Comment (pavanbalaji) [inline test/mpi/Makefile.mtest]: Changing `mtest.o` to `mtest.$(OBJEXT)` is a bug-fix and should be separated into a different commit.

File: test/mpi/Makefile_cxx.mtest [hunk 3]
```diff
@@ -12,19 +12,17 @@
 ## see Makefile.mtest for a description why this file exists, but for C++
 
 AM_CPPFLAGS = -I$(top_builddir)/include -I$(top_srcdir)/include
-LDADD = $(top_builddir)/cxx/util/mtest.o
+LDADD = $(top_builddir)/util/mtest_cxx.$(OBJEXT) $(top_builddir)/util/mtest_common.$(OBJEXT)
```
Comment (pavanbalaji) [inline test/mpi/Makefile_cxx.mtest]: Changing mtest_cxx.o to mtest_cxx.$(OBJEXT) is a bug-fix and should be separated into a different commit.

File: test/mpi/include/mpitest.h [hunk 4]
```diff
@@ -7,10 +7,10 @@
 #ifndef MPITEST_H_INCLUDED
 #define MPITEST_H_INCLUDED
 
-#include <string.h>
```
Comment (pavanbalaji) [inline test/mpi/include/mpitest.h]: Removal of unnecessary headers should be a different commit.

File: test/mpi/include/mpitestcxx.h [hunk 5]
```diff
@@ -8,28 +8,27 @@
...
+#ifdef HAVE_IOSTREAM
+#include <iostream>
+#ifdef HAVE_NAMESPACE_STD
+using namespace std;
+#endif
+#else
+#include <iostream.h>
 #endif
-
-#include <string.h>
```
Comment (pavanbalaji) [inline test/mpi/include/mpitestcxx.h]: Removal of unnecessary headers should be a different commit.

File: test/mpi/include/mpitestcxx.h [hunk 6]
```diff
@@ -8,28 +8,27 @@
...
-/*
- * Init and finalize test
- */
```
Comment (pavanbalaji) [inline test/mpi/include/mpitestcxx.h]: It's OK to remove this comment, but it should be separated into a different commit.

File: test/mpi/util/mtest.c [hunk 7]
```diff
@@ -30,7 +30,6 @@
 #endif
 #include <errno.h>
 
-
```
Comment (pavanbalaji) [inline test/mpi/util/mtest.c]: Unrelated line removal.

**Comment by hzhou** [comment]
test:jenkins/ch3/tcp

**Comment by hzhou** [comment]
@pavanbalaji I have split the commits out. I also added a patch for whitespace-checker as it does not deal with c++ header. It is ready for your another review.
